File: typescript/tests/education_board.test.ts
// education_board.test.ts
// Verifies the CircleAI.Education port: courses, lessons ordered by OrderIndex,
// enrolment + progress update, and average-progress rollup (0 when empty).

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryEducationBoard,
  EducationDomainContext,
  course,
  lesson,
  studentRecord,
} from "../src/education/index";

describe("InMemoryEducationBoard", () => {
  it("adds and retrieves courses", () => {
    const b = new InMemoryEducationBoard();
    b.addCourse(course("c1", "Algebra", "Maths", "Grade 8"));
    assert.equal(b.getCourse("c1")?.name, "Algebra");
    assert.equal(b.getCourse("nope"), undefined);
  });

  it("lists lessons ordered by OrderIndex ascending", () => {
    const b = new InMemoryEducationBoard();
    b.addLesson(lesson("l3", "c1", "Third", 600_000, 3));
    b.addLesson(lesson("l1", "c1", "First", 600_000, 1));
    b.addLesson(lesson("l2", "c1", "Second", 600_000, 2));
    b.addLesson(lesson("lx", "c2", "Other", 600_000, 1));
    assert.deepEqual(
      b.lessonsFor("c1").map((l) => l.lessonId),
      ["l1", "l2", "l3"],
    );
  });

  it("enrols students, updates progress, and computes the average", () => {
    const b = new InMemoryEducationBoard();
    b.enrol(studentRecord("s1", "A", "c1", 20));
    b.enrol(studentRecord("s2", "B", "c1", 80));
    b.enrol(studentRecord("s3", "C", "c2", 50));
    b.updateProgress("s1", 40);
    assert.equal(b.avgProgressFor("c1"), (40 + 80) / 2);
    assert.equal(b.studentsFor("c1").length, 2);
  });

  it("avgProgressFor an empty course is 0", () => {
    const b = new InMemoryEducationBoard();
    assert.equal(b.avgProgressFor("empty"), 0);
  });

  it("updateProgress for an unknown student throws", () => {
    const b = new InMemoryEducationBoard();
    assert.throws(() => b.updateProgress("ghost", 10), /Unknown student ghost/);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(EducationDomainContext.systemPromptSnippet.includes("[DOMAIN: Education]"));
    assert.deepEqual(EducationDomainContext.complianceFlags, ["SASA", "CAPS_NCS", "POPIA", "PAIA"]);
    assert.deepEqual(EducationDomainContext.suggestedTools, [
      "learning_management",
      "document_editor",
      "assessment_tools",
      "web_search",
    ]);
  });
});

File: go/tests/education_board_test.go
// education_board_test.go
//
// Verifies the CircleAI.Education port (education_board.go): course add/get,
// lesson ordering by OrderIndex, enrol + progress update, students-for filter,
// and average-progress (including the empty-course 0.0 case).

package circleai_test

import (
	"math"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestEducation_CourseAndLessons(t *testing.T) {
	b := circleai.NewInMemoryEducationBoard()
	b.AddCourse(circleai.Course{CourseId: "c1", Name: "Algebra", Subject: "Math", GradeBand: "8-9"})
	if got, ok := b.GetCourse("c1"); !ok || got.Name != "Algebra" {
		t.Fatalf("get course = %+v ok=%v", got, ok)
	}
	if _, ok := b.GetCourse("none"); ok {
		t.Fatalf("missing course found")
	}
	b.AddLesson(circleai.Lesson{LessonId: "l3", CourseId: "c1", Title: "Third", Duration: 30 * time.Minute, OrderIndex: 3})
	b.AddLesson(circleai.Lesson{LessonId: "l1", CourseId: "c1", Title: "First", Duration: 20 * time.Minute, OrderIndex: 1})
	b.AddLesson(circleai.Lesson{LessonId: "l2", CourseId: "c1", Title: "Second", Duration: 25 * time.Minute, OrderIndex: 2})
	b.AddLesson(circleai.Lesson{LessonId: "x1", CourseId: "other", Title: "Other", OrderIndex: 1})

	lessons := b.LessonsFor("c1")
	if len(lessons) != 3 || lessons[0].LessonId != "l1" || lessons[1].LessonId != "l2" || lessons[2].LessonId != "l3" {
		t.Fatalf("lessons ordered by index failed: %+v", lessons)
	}
}

func TestEducation_EnrolProgressAndAverage(t *testing.T) {
	b := circleai.NewInMemoryEducationBoard()
	b.Enrol(circleai.StudentRecord{StudentId: "s1", Name: "A", CourseId: "c1", ProgressPct: 20})
	b.Enrol(circleai.StudentRecord{StudentId: "s2", Name: "B", CourseId: "c1", ProgressPct: 40})
	b.Enrol(circleai.StudentRecord{StudentId: "s3", Name: "C", CourseId: "c2", ProgressPct: 90})

	if err := b.UpdateProgress("s1", 60); err != nil {
		t.Fatalf("update progress: %v", err)
	}
	if err := b.UpdateProgress("ghost", 10); err == nil {
		t.Fatalf("unknown student update must error")
	}
	students := b.StudentsFor("c1")
	if len(students) != 2 || students[0].StudentId != "s1" || students[1].StudentId != "s2" {
		t.Fatalf("students-for failed: %+v", students)
	}
	// Avg of 60 and 40 = 50.
	if avg := b.AvgProgressFor("c1"); math.Abs(avg-50.0) > 1e-9 {
		t.Fatalf("avg = %v, want 50", avg)
	}
	// Empty course -> 0.0.
	if avg := b.AvgProgressFor("empty"); avg != 0.0 {
		t.Fatalf("empty avg = %v, want 0", avg)
	}
}
